src/unix/process.c

```diff
@@ -326,15 +323,13 @@ static void uv__process_child_init(const uv_process_options_t* options,
 
     if (fd <= 2)
       uv__nonblock_fcntl(fd, 0);
-
-    if (close_fd >= stdio_count)
-      uv__close(close_fd);
   }
 
   for (fd = 0; fd < stdio_count; fd++) {
     use_fd = pipes[fd][1];
 
-    if (use_fd >= stdio_count)
+    if (use_fd >= stdio_count ||
+       (use_fd > STDERR_FILENO && pipes[use_fd][1] < 0))
```

Comment:
Wouldn't use_fd be O_CLOEXEC already? It seems like this wouldn't be closing `pipes[fd][0]` anymore, but I'm not sure I currently understand this part.

Nit: looks like we use 3 above instead of STDERR_FILENO, so I'd suggest doing the same here:

```suggestion
       (use_fd >= 3 && pipes[use_fd][1] < 0))
```

Reply:
@vtjnash 

> Wouldn't use_fd be O_CLOEXEC already?

For `UV_CREATE_PIPE`, both fds in `pipes[fd]` are created by `uv__make_socketpair` and have `O_CLOEXEC` set, but for `UV_INHERIT_FD` and `UV_INHERIT_STREAM`, `pipes[fd][1]` is simply copied from the original fd so it's necessary to close them.

https://github.com/libuv/libuv/blob/0fa3de1c8de2408173cb085e31dee8b82dff22ea/src/unix/process.c#L197-L202

> It seems like this wouldn't be closing pipes[fd][0] anymore

According to my understanding, they're initialized to `-1` and only re-assigned in `uv__make_socketpair`, where `O_CLOEXEC` is always set. I'm not 100% sure though so please let me know if I got anything wrong.

Reply:
> looks like we use 3 above instead of STDERR_FILENO, so I'd suggest doing the same here

I just read the code again and it seems that the `use_fd >= 3` check is actually unnecessary. If `use_fd` is in 0-2 range `pipes[use_fd][1]` will always be a valid fd because we'll open `/dev/null` even for `UV_IGNORE`:

https://github.com/libuv/libuv/blob/0fa3de1c8de2408173cb085e31dee8b82dff22ea/src/unix/process.c#L298-L312

Reply:
Ah, that seems reasonable, though I'd also have to look more closely to confirm